### ISSUE TITLE
Include URL parameters in canMatch interface

### ISSUE THREAD

**Issue by joelghill** [angular/angular#49309]
### Which @angular/* package(s) are relevant/related to the feature request?

router

### Description

The `canMatch` router guard accepts `route: Route` and `segments: UrlSegment[]` as arguments. Neither of these objects contain the parameters that may or may not be in the URL.

**Example**

A URL path will be set to `users/:userId` and when the browser navigates to the URL `users/user_123` we should have access to a mapping of parameters similar to:
```
{
    "userId": "user_123",
}
```

I can't find this information anywhere at the time the `canMatch` guard is activated. Parameters objects in the URL segments are empty, and the Route object does not have a params object, and using the activate route service also does not provide a solution. Am I doing something wrong?

### Proposed solution

Provide an object, either as a `canMatch` parameter or in the `UrlSegment` or the `Route` (I don't really care where they are, whatever makes sense to the Angular routing design) that includes the URL parameters.

### Alternatives considered

We can currently take the path in the route, and the URL segments, compare them, and derive the parameters from there. 

I feels like shouldn't have to do this though? The `canActivate` method has access to this data via the activated route snapshot (params object) and I don't understand why we can't have something similar in `canMatch`?

**Comment by atscott** [comment]
>  The canActivate method has access to this data via the activated route snapshot (params object) and I don't understand why we can't have something similar in canMatch?

This is because `canActivate` happens after matching. We don't have the parameters information until _after_ matching is finished because the route configuration affects what the parameters are.

It is possible that we could investigate accumulating a set of currently matched parameters up to the point of hitting the `Route` config and running `canMatch`. 

Note that the lack of having this information is effectively what's blocking this feature request because the experience would be sub-optimal: https://github.com/angular/angular/issues/13373

**Comment by joelghill** [comment]
> We don't have the parameters information until _after_ matching is finished because the route configuration affects what the parameters are.

Thanks for responding! I think this is the part I'm confused about. What in the route config can affect parameters at the point in time of executing `canMatch`? I'd like to understand this better.

For my use case I really need something dead simple to cover most cases:

1. Route path is `something/:something_id` -> we currently have this
2. When we navigate to a URL that matches the above path, Url Segments are `['something', '123456']` -> we currently have these
3. Parameters are `{'something': '123456'}` -> this is what I need to perform the `canMatch`

Unless I'm missing something I could probably write a custom guard that does the parsing I need, I'm just surprised it's not there already and I'm unaware of what in the overall architecture would prevent this. Again, I don't have a deep understanding of what angular is doing under the hood though.

**Comment by joelghill** [comment]
I should also add that perhaps `canActivate` is a better solution in my use case, but it does not come with the benefit of automatically trying a different pattern on failure. Being able to setup a wildcard route with a generic "resource not found" component when none of the routes match is a great feature but I find we really can't replace `canActivate` with `canMatch` with it's current interface.

**Comment by atscott** [comment]
> Route path is `something/:something_id` -> we currently have this

We absolutely have an easy way to access the params of the _current_ route being matched. However these are not final because a redirect may cause parameters and parameter names to change. You _can_ get this information manually by just rerunning the matcher on those segments in your guard: 

```
canMatch: [
          (route: Route, segments: UrlSegment[]) => {
            const matcher= route.matcher || defaultUrlMatcher;
            const matchResult = matcher(
              segments,
              new UrlSegmentGroup(segments, {}),
              route
            );
            console.log(matchResult);
            return true;
          },
        ],
```
https://stackblitz.com/edit/angular-6m5gvr?file=src%2Fmain.ts

These are just the current route though and again as I mentioned above, we don't currently pass down the parameter matches from the parent configs (you also would definitely not have access to any child parameters no mater what). Also, because we haven't matched all the parent's children, we can't construct the parent tree object (`UrlSegmentGroup` or `ActivatedRoute`) until this is done.

>  Parameters are {'something': '123456'} -> this is what I need to perform the canMatch
...
guard accepts route: Route and segments: UrlSegment[] as arguments. Neither of these objects contain the parameters that may or may not be in the URL.

I think part of the confusion is that the `parameters` on `UrlSegment` are _not_ the path parameter names in the route config. These would be the matrix parameters like in a URL `/a;something_id=1234`.

Long story short, I _do_ think this is a problem that we should fix, especially if we want to land https://github.com/angular/angular/issues/13373 (and we do). I would guess that nobody really thought about this before.

**Comment by joelghill** [comment]
Thank you @atscott, your response and sample code was super helpful! 

I was eventually able to get done what I needed to with the `canMatch` guard but I think there is more information that could be surfaced to the developer at the time the guard is executed.

**Comment by matheo** [comment]
@atscott nice snippet to parse the current params, but I was thinking we could know the params parsed for parent routes and it seems not too easy right? injecting the `Router` and going through the current state seems like the only option so far.

**Comment by skwny** [comment]
CanMatch would be a lot more useful with router state.
